File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/Shared/LayerStates.vb
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices

Public Class LayerStates
    'CType(LayerStateMasks.Color + LayerStateMasks.LineType, LayerStateMasks)
    Public Shared Function Create(doc As Document, layerStateName As String, layerStateMasks As LayerStateMasks) _
        As Boolean
        Try
            Using lock As DocumentLock = doc.LockDocument
                Dim layerStateManager As LayerStateManager = doc.Database.LayerStateManager
                Dim tempLayerStateName As String = ""
                Dim leaveMissingLayersUnchanged As Integer = 0
                If layerStateManager.HasLayerState(layerStateName) = True Then
                    tempLayerStateName = CreateFirstAvailableLayerStateName(doc, "tempLayerState", layerStateMasks)
                    layerStateManager.RestoreLayerState(tempLayerStateName,
                        ObjectId.Null,
                        leaveMissingLayersUnchanged,
                        layerStateMasks)
                    layerStateManager.DeleteLayerState(layerStateName)
                End If
                layerStateManager.SaveLayerState(layerStateName, layerStateMasks, ObjectId.Null)
                layerStateManager.RestoreLayerState(layerStateName,
                    ObjectId.Null,
                    leaveMissingLayersUnchanged,
                    layerStateMasks)
                If Not String.IsNullOrEmpty(tempLayerStateName) Then
                    layerStateManager.DeleteLayerState(tempLayerStateName)
                End If
            End Using
        Catch ex As Exception
            Throw (ex)
        End Try
        Return True
    End Function

    Public Shared Function CreateFirstAvailableLayerStateName(doc As Document,
        layerStateName As String,
        layerStateMasks As LayerStateMasks) As String
        Dim i As Integer = 0
        Dim currentLayerStateName As String = layerStateName
        Try
            Using lock As DocumentLock = doc.LockDocument
                Dim layerStateManager As LayerStateManager = doc.Database.LayerStateManager
                Dim layerStateExists As Boolean = layerStateManager.HasLayerState(layerStateName)
                While layerStateExists
                    i = i + 1
                    currentLayerStateName = layerStateName + "(" + i.ToString + ")"
                    layerStateExists = layerStateManager.HasLayerState(currentLayerStateName)
                End While
                layerStateManager.SaveLayerState(layerStateName, layerStateMasks, ObjectId.Null)
                Return currentLayerStateName
            End Using
        Catch ex As Exception
            Throw (ex)
        End Try
        Return Nothing
    End Function


    Public Shared Function GetLastRestoredLayerState(doc As Document) As String
        Dim i As Integer = 0
        Try
            Using lock As DocumentLock = doc.LockDocument
                Dim layerStateManager As LayerStateManager = doc.Database.LayerStateManager
                Return layerStateManager.LastRestoredLayerState
            End Using
        Catch ex As Exception
            Throw (ex)
        End Try
        Return Nothing
    End Function

    Public Shared Function Restore(doc As Document, layerState As String, layerStateMasks As LayerStateMasks) As Boolean
        Dim i As Integer = 0
        Try
            Using lock As DocumentLock = doc.LockDocument
                Dim layerStateManager As LayerStateManager = doc.Database.LayerStateManager
                If layerStateManager.HasLayerState(layerState) Then
                    Dim leaveMissingLayersUnchanged As Integer = 0
                    layerStateManager.RestoreLayerState(layerState,
                        ObjectId.Null,
                        leaveMissingLayersUnchanged,
                        layerStateMasks)
                    Return True
                End If
            End Using
        Catch ex As Exception
            Throw (ex)
        End Try
        Return False
    End Function
End Class

File: cadwiki-nuget/cadwiki.AC/Utilities/LayerStates.cs
using System;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.DatabaseServices;

namespace cadwiki.AC.Utilities
{

    public class LayerStates
    {
        // CType(LayerStateMasks.Color + LayerStateMasks.LineType, LayerStateMasks)
        public static bool Create(Document doc, string layerStateName, LayerStateMasks layerStateMasks)
        {
            try
            {
                using (var @lock = doc.LockDocument())
                {
                    var layerStateManager = doc.Database.LayerStateManager;
                    string tempLayerStateName = "";
                    int leaveMissingLayersUnchanged = 0;
                    if (layerStateManager.HasLayerState(layerStateName) == true)
                    {
                        tempLayerStateName = CreateFirstAvailableLayerStateName(doc, "tempLayerState", layerStateMasks);
                        layerStateManager.RestoreLayerState(tempLayerStateName, ObjectId.Null, leaveMissingLayersUnchanged, layerStateMasks);
                        layerStateManager.DeleteLayerState(layerStateName);
                    }
                    layerStateManager.SaveLayerState(layerStateName, layerStateMasks, ObjectId.Null);
                    layerStateManager.RestoreLayerState(layerStateName, ObjectId.Null, leaveMissingLayersUnchanged, layerStateMasks);
                    if (!string.IsNullOrEmpty(tempLayerStateName))
                    {
                        layerStateManager.DeleteLayerState(tempLayerStateName);
                    }
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return true;
        }

        public static string CreateFirstAvailableLayerStateName(Document doc, string layerStateName, LayerStateMasks layerStateMasks)
        {
            int i = 0;
            string currentLayerStateName = layerStateName;
            try
            {
                using (var @lock = doc.LockDocument())
                {
                    var layerStateManager = doc.Database.LayerStateManager;
                    bool layerStateExists = layerStateManager.HasLayerState(layerStateName);
                    while (layerStateExists)
                    {
                        i = i + 1;
                        currentLayerStateName = layerStateName + "(" + i.ToString() + ")";
                        layerStateExists = layerStateManager.HasLayerState(currentLayerStateName);
                    }
                    layerStateManager.SaveLayerState(layerStateName, layerStateMasks, ObjectId.Null);
                    return currentLayerStateName;
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return null;
        }


        public static string GetLastRestoredLayerState(Document doc)
        {
            int i = 0;
            try
            {
                using (var @lock = doc.LockDocument())
                {
                    var layerStateManager = doc.Database.LayerStateManager;
                    return layerStateManager.LastRestoredLayerState;
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return null;
        }

        public static bool Restore(Document doc, string layerState, LayerStateMasks layerStateMasks)
        {
            int i = 0;
            try
            {
                using (var @lock = doc.LockDocument())
                {
                    var layerStateManager = doc.Database.LayerStateManager;
                    if (layerStateManager.HasLayerState(layerState))
                    {
                        int leaveMissingLayersUnchanged = 0;
                        layerStateManager.RestoreLayerState(layerState, ObjectId.Null, leaveMissingLayersUnchanged, layerStateMasks);
                        return true;
                    }
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return false;
        }
    }
}